### ISSUE TITLE
Rule proposal: A rule checks optional chaining in logical expression

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1801]
### Description

For "logical or", the optional chaining should be consistent.
For "logical and", should not use unnecessary `?.`(This may makes code harder to edit, need discussion).

### Fail

```js
foo?.bar || foo.baz;
// The right side should check foo
```

```js
foo.bar || foo?.baz;
// The right side should not check foo
```

```js
foo?.bar && foo?.baz;
// The right side `?.` is useless
```

```js
foo.bar && foo?.baz;
// The right side `?.` is useless
```

### Pass

```js
foo?.bar || foo?.baz;
```

```js
foo.bar || foo.baz;
```

```js
foo?.bar && foo.baz;
```

```js
foo.bar && foo.baz;
```

**Comment by sindresorhus** [comment]
Name suggestions for the rule?

**Comment by sindresorhus** [comment]
`consistent-optional-chaining`?

**Comment by sindresorhus** [comment]
Accepted.

Let's go with the name `consistent-optional-chaining`.

**Comment by Samuel-Therrien-Beslogic** [comment]
Copied from https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2230

I'd like to catch the following cases:

### Fail

```js
// Nullish coalescing
foo == null ? bar : foo // Autofix to `foo ?? bar`
foo == undefined ? bar : foo // Autofix to `foo ?? bar`
foo == false ? bar : foo // Same suggestion as `!foo ? bar : foo`
foo == true ? foo : bar // Same suggestion as `foo ? foo : bar`
// Optional chaining
foo === undefined ? foo : foo.bar // Autofix to `foo?.bar`
foo == undefined ? undefined : foo.bar // Autofix to `foo?.bar`
foo == null ? undefined : foo.bar // Autofix to `foo?.bar`
```

### Pass

```js
foo == null ? foo : bar
foo == undefined ? foo : bar
foo == false ? foo : bar
foo == true ? bar : foo
foo == false ? bar : false
foo == true ? true : bar
// Technically these can change the return type from `null` to `undefined`. So I'm not sure if they should pass or fail with suggestion
foo == undefined ? foo : foo.bar // Suggests `foo?.bar`
foo == null ? foo : foo.bar // Suggests `foo?.bar`
```

### Additional Info

https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-redundant-boolean.md catches the redundant falsy/truthy boolean conditional in a ternary. But offers no autofix.

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v46.0.1/docs/rules/no-negated-condition.md already takes cares of the `!=` cases to I didn't list them here.

`ts(5076)` prevents mixing `&&`/`||` with `??` without parenthesis. So some autofix might require adding parenthesis.

**Comment by andersk** [comment]
> ```js
> foo == false ? bar : foo // Same suggestion as `!foo ? bar : foo`
> foo == true ? foo : bar // Same suggestion as `foo ? foo : bar`
> ```

Not equivalent when `foo` is `undefined` or `null` or `[]` (and optional configuration flags of type `boolean | undefined` aren’t uncommon).

> ```js
> foo === undefined ? foo : foo.bar // Autofix to `foo?.bar`
> ```

Not equivalent when `foo` is `null`.

**Comment by silverwind** [comment]
Another case with function:

Fail:

```js
if (foo) foo();
```

Pass:

```js
foo?.();
```
